LanguageFeatures/Private-fields-promotion/promotion_A02_t02.dart

```diff
@@ -22,15 +22,21 @@ class A {
   void foo() {}
 }
 
-class C<T extends A> {
+class C<T extends A?> {
   late final T? _x;
 
   void test(T t) {
     _x = t;
-    _x.foo();
+    if (_x != null) {
+      _x.foo();
```

Comment:
This one is questionable: `_x` had the type `T?` which is not a type variable, which means that we can't promote it to an intersection type. Also, `Object` is not a subtype of `T?`, so we can't just promote it to `Object`. So we should not get a promotion here.

Anyway, we probably do get that promotion, because both the analyzer and the CFE treat `T?` as a representation of `T` where the nullability is "nullable", not as a type operator which is applied to `T`.

@stereotype441, what's going on with the issues linked from https://github.com/dart-lang/language/pull/2427?

Reply:
@eernstg 
I think this is slightly different from what's going on in https://github.com/dart-lang/language/pull/2427. That PR is about the behaviour of UP, which isn't involved here.

What's happening here is that the null check causes flow analysis to invoke the **NonNull** function on the type `T?`. The relevant rules for **NonNull** are:

> **NonNull**(_C_<_T_<sub>1</sub>, ... , _T_<sub>_n_</sub>>) = _C_<_T_<sub>1</sub>, ... , _T_<sub>_n_</sub>>  for class *C* other than Null (including Object).
> **NonNull**(_X_) = X & **NonNull**(B), where B is the bound of X.
> **NonNull**(_T_?) = **NonNull**(_T_)

So, applying those rules, **NonNull**(`T?`) = **NonNull**(`T`) = `T & `**NonNull**(`A?`) = `T & `**NonNull**(`A`) = `T & A`

And it's valid to call `.foo` on the type `T & A`. So I think this test case is just fine.

As to your other question about what's going on with the issues linked from https://github.com/dart-lang/language/pull/2427, I see just one issue, https://github.com/dart-lang/sdk/issues/49691, and it's assigned to @chloestefantsova, so she's probably in the best position to answer that.

Reply:
No changes here